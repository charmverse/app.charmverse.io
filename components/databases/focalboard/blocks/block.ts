// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import difference from 'lodash/difference'

import {Utils} from '../utils'

const contentBlockTypes = ['text', 'image', 'divider', 'checkbox'] as const
const blockTypes = [...contentBlockTypes, 'board', 'view', 'card', 'comment', 'unknown'] as const
type ContentBlockTypes = typeof contentBlockTypes[number]
type BlockTypes = typeof blockTypes[number]

interface BlockPatch {
    workspaceId?: string
    parentId?: string
    rootId?: string
    schema?: number
    type?: BlockTypes
    title?: string
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    updatedFields?: Record<string, any>
    deletedFields?: string[]
    deleteAt?: number
}

interface Block {
    id: string
    workspaceId: string
    parentId: string
    rootId: string
    createdBy: string
    modifiedBy: string

    schema: number
    type: BlockTypes
    title: string
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    fields: Record<string, any>

    createAt: number
    updateAt: number
    deleteAt: number
}

function createBlock(block?: Block): Block {
    const now = Date.now()
    return {
        id: block?.id || Utils.createGuid(Utils.blockTypeToIDType(block?.type)),
        schema: 1,
        workspaceId: block?.workspaceId || '',
        parentId: block?.parentId || '',
        rootId: block?.rootId || '',
        createdBy: block?.createdBy || '',
        modifiedBy: block?.modifiedBy || '',
        type: block?.type || 'unknown',
        fields: block?.fields ? {...block?.fields} : {},
        title: block?.title || '',
        createAt: block?.createAt || now,
        updateAt: block?.updateAt || now,
        deleteAt: block?.deleteAt || 0,
    }
}

// createPatchesFromBlock creates two BlockPatch instances, one that
// contains the delta to update the block and another one for the undo
// action, in case it happens
function createPatchesFromBlocks(newBlock: Block, oldBlock: Block): BlockPatch[] {
    const newDeletedFields = difference(Object.keys(newBlock.fields), Object.keys(oldBlock.fields))
    const newUpdatedFields: Record<string, any> = {}
    const newUpdatedData: Record<string, any> = {}
    Object.keys(newBlock.fields).forEach((val) => {
        if (oldBlock.fields[val] !== newBlock.fields[val]) {
            newUpdatedFields[val] = newBlock.fields[val]
        }
    })
    Object.keys(newBlock).forEach((val) => {
        if (val !== 'fields' && (oldBlock as any)[val] !== (newBlock as any)[val]) {
            newUpdatedData[val] = (newBlock as any)[val]
        }
    })

    const oldDeletedFields = difference(Object.keys(oldBlock.fields), Object.keys(newBlock.fields))
    const oldUpdatedFields: Record<string, any> = {}
    const oldUpdatedData: Record<string, any> = {}
    Object.keys(oldBlock.fields).forEach((val) => {
        if (oldBlock.fields[val] !== newBlock.fields[val]) {
            oldUpdatedFields[val] = oldBlock.fields[val]
        }
    })
    Object.keys(oldBlock).forEach((val) => {
        if (val !== 'fields' && (oldBlock as any)[val] !== (newBlock as any)[val]) {
            oldUpdatedData[val] = (oldBlock as any)[val]
        }
    })

    return [
        {
            ...newUpdatedData,
            updatedFields: newUpdatedFields,
            deletedFields: oldDeletedFields,
        },
        {
            ...oldUpdatedData,
            updatedFields: oldUpdatedFields,
            deletedFields: newDeletedFields,
        },
    ]
}

export type {ContentBlockTypes, BlockTypes}
export {blockTypes, contentBlockTypes, Block, BlockPatch, createBlock, createPatchesFromBlocks}
