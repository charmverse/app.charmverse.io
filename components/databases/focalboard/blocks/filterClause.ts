// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils} from '../utils'

type FilterCondition = 'includes' | 'notIncludes' | 'isEmpty' | 'isNotEmpty'

type FilterClause = {
    propertyId: string
    condition: FilterCondition
    values: string[]
}

function createFilterClause(o?: FilterClause): FilterClause {
    return {
        propertyId: o?.propertyId || '',
        condition: o?.condition || 'includes',
        values: o?.values?.slice() || [],
    }
}

function areEqual(a: FilterClause, b: FilterClause): boolean {
    return (
        a.propertyId === b.propertyId &&
        a.condition === b.condition &&
        Utils.arraysEqual(a.values, b.values)
    )
}

export {FilterClause, FilterCondition, createFilterClause, areEqual}
