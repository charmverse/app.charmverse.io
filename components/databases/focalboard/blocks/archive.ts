// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Block} from './block'

interface ArchiveHeader {
    version: number
    date: number
}

interface ArchiveLine {
    type: string,
    data: unknown,
}

// This schema allows the expansion of additional line types in the future
interface BlockArchiveLine extends ArchiveLine {
    type: 'block',
    data: Block
}

class ArchiveUtils {
    static buildBlockArchive(blocks: readonly Block[]): string {
        const header: ArchiveHeader = {
            version: 1,
            date: Date.now(),
        }

        const headerString = JSON.stringify(header)
        let content = headerString + '\n'
        for (const block of blocks) {
            const line: BlockArchiveLine = {
                type: 'block',
                data: block,
            }
            const lineString = JSON.stringify(line)
            content += lineString
            content += '\n'
        }

        return content
    }

    static parseBlockArchive(contents: string): Block[] {
        const blocks: Block[] = []
        const allLineStrings = contents.split('\n')
        if (allLineStrings.length >= 2) {
            const headerString = allLineStrings[0]
            const header = JSON.parse(headerString) as ArchiveHeader
            if (header.date && header.version >= 1) {
                const lineStrings = allLineStrings.slice(1)
                let lineNum = 2
                for (const lineString of lineStrings) {
                    if (!lineString) {
                        // Ignore empty lines, e.g. last line
                        continue
                    }
                    const line = JSON.parse(lineString) as ArchiveLine
                    if (!line || !line.type || !line.data) {
                        throw new Error(`ERROR parsing line ${lineNum}`)
                    }
                    switch (line.type) {
                    case 'block': {
                        const blockLine = line as BlockArchiveLine
                        const block = blockLine.data
                        blocks.push(block)
                        break
                    }
                    }

                    lineNum += 1
                }
            } else {
                throw new Error('ERROR parsing header')
            }
        }

        return blocks
    }
}

export {ArchiveHeader, ArchiveLine, BlockArchiveLine, ArchiveUtils}
