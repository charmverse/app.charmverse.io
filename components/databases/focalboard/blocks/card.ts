// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Block, createBlock} from './block'

type CardFields = {
    icon?: string
    isTemplate?: boolean
    properties: Record<string, string | string[]>
    contentOrder: Array<string | string[]>
}

type Card = Block & {
    fields: CardFields
}

function createCard(block?: Block): Card {
    const contentOrder: Array<string|string[]> = []
    const contentIds = block?.fields?.contentOrder?.filter((id: any) => id !== null)

    if (contentIds?.length > 0) {
        for (const contentId of contentIds) {
            if (typeof contentId === 'string') {
                contentOrder.push(contentId)
            } else {
                contentOrder.push(contentId.slice())
            }
        }
    }
    return {
        ...createBlock(block),
        type: 'card',
        fields: {
            icon: block?.fields.icon || '',
            properties: {...(block?.fields.properties || {})},
            contentOrder,
            isTemplate: block?.fields.isTemplate || false,
        },
    }
}

export {Card, createCard}
