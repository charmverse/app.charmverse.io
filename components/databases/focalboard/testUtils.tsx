// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {IntlProvider} from 'react-intl'
import React from 'react'
import {DndProvider} from 'react-dnd'
import {HTML5Backend} from 'react-dnd-html5-backend'
import configureStore, {MockStoreEnhanced} from 'redux-mock-store'
import {Middleware} from 'redux'

import {Block} from './blocks/block'

export const wrapIntl = (children?: React.ReactNode): JSX.Element => <IntlProvider locale='en'>{children}</IntlProvider>
export const wrapDNDIntl = (children?: React.ReactNode): JSX.Element => {
    return (
        <DndProvider backend={HTML5Backend}>
            {wrapIntl(children)}
        </DndProvider>
    )
}

export function mockDOM(): void {
    window.focus = jest.fn()
    document.createRange = () => {
        const range = new Range()
        range.getBoundingClientRect = jest.fn()
        range.getClientRects = () => {
            return {
                item: () => null,
                length: 0,
                [Symbol.iterator]: jest.fn(),
            }
        }
        return range
    }
}
export function mockMatchMedia(result: {matches: boolean}): void {
    // We check if system preference is dark or light theme.
    // This is required to provide it's definition since
    // window.matchMedia doesn't exist in Jest.
    Object.defineProperty(window, 'matchMedia', {
        writable: true,
        value: jest.fn().mockImplementation(() => {
            return result

            // return ({
            //     matches: true,
            // })
        }),
    })
}

export function mockStateStore(middleware:Middleware[], state:unknown): MockStoreEnhanced<unknown, unknown> {
    const mockStore = configureStore(middleware)
    return mockStore(state)
}

export type BlocksById<BlockType> = {[key: string]: BlockType}

export function blocksById<BlockType extends Block>(blocks: Array<BlockType>): BlocksById<BlockType> {
    return blocks.reduce((res, block) => {
        res[block.id] = block
        return res
    }, {} as BlocksById<BlockType>)
}
