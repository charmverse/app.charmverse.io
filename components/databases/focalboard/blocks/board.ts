// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.
import {Utils, IDType} from '../utils'

import {Block, createBlock} from './block'
import {Card} from './card'

type PropertyType = 'text' | 'number' | 'select' | 'multiSelect' | 'date' | 'person' | 'file' | 'checkbox' | 'url' | 'email' | 'phone' | 'createdTime' | 'createdBy' | 'updatedTime' | 'updatedBy'

interface IPropertyOption {
    id: string
    value: string
    color: string
}

// A template for card properties attached to a board
interface IPropertyTemplate {
    id: string
    name: string
    type: PropertyType
    options: IPropertyOption[]
}

type BoardFields = {
    icon: string
    description: string
    showDescription?: boolean
    isTemplate?: boolean
    cardProperties: IPropertyTemplate[]
    columnCalculations: Record<string, string>
}

type Board = Block & {
    fields: BoardFields
}

function createBoard(block?: Block): Board {
    let cardProperties: IPropertyTemplate[] = []
    const selectProperties = cardProperties.find((o) => o.type === 'select')
    if (!selectProperties) {
        const property: IPropertyTemplate = {
            id: Utils.createGuid(IDType.BlockID),
            name: 'Status',
            type: 'select',
            options: [],
        }
        cardProperties.push(property)
    }

    if (block?.fields.cardProperties) {
        // Deep clone of card properties and their options
        cardProperties = block?.fields.cardProperties.map((o: IPropertyTemplate) => {
            return {
                id: o.id,
                name: o.name,
                type: o.type,
                options: o.options ? o.options.map((option) => ({...option})) : [],
            }
        })
    }

    return {
        ...createBlock(block),
        type: 'board',
        fields: {
            showDescription: block?.fields.showDescription || false,
            description: block?.fields.description || '',
            icon: block?.fields.icon || '',
            isTemplate: block?.fields.isTemplate || false,
            columnCalculations: block?.fields.columnCalculations || [],
            cardProperties,
        },
    }
}

type BoardGroup = {
    option: IPropertyOption
    cards: Card[]
}

export {Board, PropertyType, IPropertyOption, IPropertyTemplate, BoardGroup, createBoard}
