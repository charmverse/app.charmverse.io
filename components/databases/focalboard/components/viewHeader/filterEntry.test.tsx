// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react'
import {render, screen} from '@testing-library/react'
import {Provider as ReduxProvider} from 'react-redux'

import '@testing-library/jest-dom'
import userEvent from '@testing-library/user-event'

import {mocked} from 'ts-jest/utils'

import {FilterClause} from '../../blocks/filterClause'

import {TestBlockFactory} from '../../test/testBlockFactory'

import {wrapIntl, mockStateStore} from '../../testUtils'

import mutator from '../../mutator'

import FilterEntry from './filterEntry'

jest.mock('../../mutator')
const mockedMutator = mocked(mutator, true)

const board = TestBlockFactory.createBoard()
const activeView = TestBlockFactory.createBoardView(board)
const filter: FilterClause = {
    propertyId: '1',
    condition: 'includes',
    values: ['Status'],
}
const state = {
    users: {
        me: {
            id: 'user-id-1',
            username: 'username_1',
        },
    },
}
const store = mockStateStore([], state)
const mockedConditionClicked = jest.fn()

describe('components/viewHeader/filterEntry', () => {
    beforeEach(() => {
        jest.clearAllMocks()
        board.fields.cardProperties[0].options = [{id: 'Status', value: 'Status', color: ''}]
        activeView.fields.filter.filters = [filter]
    })
    test('return filterEntry', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[0]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
    })

    test('return filterEntry and click on status', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[0]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
        const buttonStatus = screen.getByRole('button', {name: 'Status'})
        userEvent.click(buttonStatus)
        expect(mockedMutator.changeViewFilter).toBeCalledTimes(1)
    })
    test('return filterEntry and click on includes', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[1]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
        const buttonIncludes = screen.getAllByRole('button', {name: 'includes'})[1]
        userEvent.click(buttonIncludes)
        expect(mockedConditionClicked).toBeCalledTimes(1)
    })
    test('return filterEntry and click on doesn\'t include', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[1]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
        const buttonNotInclude = screen.getByRole('button', {name: 'doesn\'t include'})
        userEvent.click(buttonNotInclude)
        expect(mockedConditionClicked).toBeCalledTimes(1)
    })
    test('return filterEntry and click on is empty', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[1]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
        const buttonEmpty = screen.getByRole('button', {name: 'is empty'})
        userEvent.click(buttonEmpty)
        expect(mockedConditionClicked).toBeCalledTimes(1)
    })
    test('return filterEntry and click on is not empty', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[1]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
        const buttonNotEmpty = screen.getByRole('button', {name: 'is not empty'})
        userEvent.click(buttonNotEmpty)
        expect(mockedConditionClicked).toBeCalledTimes(1)
    })
    test('return filterEntry and click on delete', () => {
        const {container} = render(
            wrapIntl(
                <ReduxProvider store={store}>
                    <FilterEntry
                        board={board}
                        view={activeView}
                        conditionClicked={mockedConditionClicked}
                        filter={filter}
                    />
                </ReduxProvider>,
            ),
        )
        const buttonElement = screen.getAllByRole('button', {name: 'menuwrapper'})[1]
        userEvent.click(buttonElement)
        expect(container).toMatchSnapshot()
        const allButton = screen.getAllByRole('button')
        userEvent.click(allButton[allButton.length - 1])
        expect(mockedMutator.changeViewFilter).toBeCalledTimes(1)
    })
})
