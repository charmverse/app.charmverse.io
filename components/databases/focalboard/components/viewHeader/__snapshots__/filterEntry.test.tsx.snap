// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/viewHeader/filterEntry return filterEntry 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="Status"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Status
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="Property 1"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Property 1
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="Property 2"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Property 2
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="Property 3"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Property 3
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;

exports[`components/viewHeader/filterEntry return filterEntry and click on delete 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="includes"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                includes
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="doesn't include"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                doesn't include
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is empty
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is not empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is not empty
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;

exports[`components/viewHeader/filterEntry return filterEntry and click on doesn't include 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="includes"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                includes
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="doesn't include"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                doesn't include
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is empty
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is not empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is not empty
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;

exports[`components/viewHeader/filterEntry return filterEntry and click on includes 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="includes"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                includes
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="doesn't include"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                doesn't include
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is empty
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is not empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is not empty
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;

exports[`components/viewHeader/filterEntry return filterEntry and click on is empty 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="includes"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                includes
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="doesn't include"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                doesn't include
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is empty
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is not empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is not empty
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;

exports[`components/viewHeader/filterEntry return filterEntry and click on is not empty 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="includes"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                includes
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="doesn't include"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                doesn't include
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is empty
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="is not empty"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                is not empty
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;

exports[`components/viewHeader/filterEntry return filterEntry and click on status 1`] = `
<div>
  <div
    class="FilterEntry"
  >
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          (unknown)
        </span>
      </button>
      <div
        class="Menu noselect bottom"
      >
        <div
          class="menu-contents"
        >
          <div
            class="menu-options"
          >
            <div
              aria-label="Status"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Status
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="Property 1"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Property 1
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="Property 2"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Property 2
              </div>
              <div
                class="noicon"
              />
            </div>
            <div
              aria-label="Property 3"
              class="MenuOption TextOption menu-option"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Property 3
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
          <div
            class="menu-spacer hideOnWidescreen"
          />
          <div
            class="menu-options hideOnWidescreen"
          >
            <div
              aria-label="Cancel"
              class="MenuOption TextOption menu-option menu-cancel"
              role="button"
            >
              <div
                class="noicon"
              />
              <div
                class="menu-name"
              >
                Cancel
              </div>
              <div
                class="noicon"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-label="menuwrapper"
      class="MenuWrapper"
      role="button"
    >
      <button
        class="Button"
        type="button"
      >
        <span>
          includes
        </span>
      </button>
    </div>
    <div
      class="octo-spacer"
    />
    <button
      class="Button"
      type="button"
    >
      <span>
        Delete
      </span>
    </button>
  </div>
</div>
`;
